token-metadata/program/src/processor/metadata/update.rs

```diff
@@ -197,60 +251,136 @@ fn validate_update(
     args: &UpdateArgs,
     authority_type: &AuthorityType,
     metadata_delegate_role: Option<MetadataDelegateRole>,
+    desired_token_standard: Option<TokenStandard>,
 ) -> ProgramResult {
     // validate the authority type
-
     match authority_type {
         AuthorityType::Metadata => {
-            // metadata authority is the paramount (upadte) authority
+            // metadata authority is the paramount (update) authority
             msg!("Auth type: Metadata");
         }
-        AuthorityType::MetadataDelegate => {
-            // support for delegate update
-            msg!("Auth type: Delegate");
-        }
         AuthorityType::Holder => {
             // support for holder update
             msg!("Auth type: Holder");
             return Err(MetadataError::FeatureNotSupported.into());
         }
-        _ => {
-            return Err(MetadataError::InvalidAuthorityType.into());
+        AuthorityType::MetadataDelegate => {
+            // support for delegate update
+            msg!("Auth type: Delegate");
         }
+        _ => return Err(MetadataError::InvalidAuthorityType.into()),
     }
 
-    let UpdateArgs::V1 {
+    // Destructure args.
+    let (
+        new_update_authority,
         data,
         primary_sale_happened,
         is_mutable,
         collection,
+        collection_details,
         uses,
+        rule_set,
+    ) = get_update_args_fields!(
+        args,
         new_update_authority,
+        data,
+        primary_sale_happened,
+        is_mutable,
+        collection,
         collection_details,
-        ..
-    } = args;
+        uses,
+        rule_set
+    );
 
     // validate the delegate role: this consist in checking that
     // the delegate is only updating fields that it has access to
-    match metadata_delegate_role {
-        Some(MetadataDelegateRole::ProgrammableConfig) => {
-            // can only update the programmable config
-            if data.is_some()
-                || primary_sale_happened.is_some()
-                || is_mutable.is_some()
-                || collection.is_some()
-                || uses.is_some()
-                || new_update_authority.is_some()
-                || collection_details.is_some()
-            {
-                return Err(MetadataError::InvalidUpdateArgs.into());
+    if let Some(metadata_delegate_role) = metadata_delegate_role {
+        match metadata_delegate_role {
+            MetadataDelegateRole::Authority => {
+                // Fields allowed for `Authority`:
+                // `new_update_authority`
+                // `primary_sale_happened`
+                // `is_mutable`
+                // `token_standard`
+                if data.is_some()
```

Comment:
The main concept that stood out with @samuelvanderwaal 's solution above is the general concept of using type safety.  Iterated on the idea above and eventually landed on the idea to use different `UpdateArgs` struct types for different delegates.  This results in a bit of extra struct type handling, but the end result is that it is 100% clear to the user from the `UpdateArgs` what exact fields a delegate is allowed to pass into `Update`.  It is not necessary to read the `Update` code or consult docs to find out this information.

Discussed with team and made the changes to this PR.  Also @lorisleiva mentioned that different handlers for the different delegates can actually be generated for the JS SDK making it easier for a delegate to use.

Reply:
So now UpdateArgs looks like this:
```Rust
pub enum UpdateArgs {
    V1 {
        /// The new update authority.
        new_update_authority: Option<Pubkey>,
        /// The metadata details.
        data: Option<Data>,
        /// Indicates whether the primary sale has happened or not (once set to `true`, it cannot be
        /// changed back).
        primary_sale_happened: Option<bool>,
        // Indicates Whether the data struct is mutable or not (once set to `true`, it cannot be
        /// changed back).
        is_mutable: Option<bool>,
        /// Collection information.
        collection: CollectionToggle,
        /// Additional details of the collection.
        collection_details: CollectionDetailsToggle,
        /// Uses information.
        uses: UsesToggle,
        // Programmable rule set configuration (only applicable to `Programmable` asset types).
        rule_set: RuleSetToggle,
        /// Required authorization data to validate the request.
        authorization_data: Option<AuthorizationData>,
    },
    UpdateAuthorityV2(UpdateAuthorityV2),
    AuthorityItemDelegateV2(AuthorityItemDelegateV2),
    CollectionDelegateV2(CollectionDelegateV2),
    DataDelegateV2(DataDelegateV2),
    ProgConfigDelegateV2(ProgConfigDelegateV2),
    DataItemDelegateV2(DataItemDelegateV2),
    CollectionItemDelegateV2(CollectionItemDelegateV2),
    ProgConfigItemDelegateV2(ProgConfigItemDelegateV2),
}
```
And the individual structs look like this:
```Rust
pub struct AuthorityItemDelegateV2 {
    /// The new update authority.
    pub new_update_authority: Option<Pubkey>,
    /// Indicates whether the primary sale has happened or not (once set to `true`, it cannot be
    /// changed back).
    pub primary_sale_happened: Option<bool>,
    // Indicates Whether the data struct is mutable or not (once set to `true`, it cannot be
    /// changed back).
    pub is_mutable: Option<bool>,
    /// Token standard.
    pub token_standard: Option<TokenStandard>,
}
```

Rather than struct-like enum variants, I used tuple enum variants backed by actual struct types.  This is so that I can implement the `Default` trait separately for each struct and thus each variant. This in turn enables more ergonomic Rust handling such as instantiating with only setting the field I care about and default for the remainder:
```Rust
let update_args = UpdateArgs::from(UpdateAuthorityV2 {
    rule_set: RuleSetToggle::Clear,
    ..Default::default()
});
```